contracts/src/systems/place_card.cairo

```diff
@@ -23,27 +23,72 @@ mod place_card_system {
     fn execute(ctx: Context, game_id: felt252, card_id: u256, position: Roles) {
         let game_player_key: (felt252, felt252) = (game_id, ctx.origin.into());
         let mut player = get!(ctx.world, game_player_key, Player);
-        let card = get!(ctx.world, (card_id), Card);
+        let mut card = get!(ctx.world, (card_id), Card);
         assert(player.remaining_energy >= card.cost.into(), 'Not enough energy');
+
+        let mut is_on_its_role = false;
         match position {
             Roles::Goalkeeper => {
                 assert(player.goalkeeper.is_none(), 'Goalkeeper already placed');
                 player.goalkeeper = Option::Some(Placement::Side(card_id));
+                match card.role {
+                    Roles::Goalkeeper => {
+                        is_on_its_role = true;
+                    },
+                    Roles::Defender => (),
+                    Roles::Midfielder => (),
+                    Roles::Attacker => (),
+                }
             },
             Roles::Defender => {
                 assert(player.defender.is_none(), 'Defender already placed');
                 player.defender = Option::Some(Placement::Side(card_id));
+                match card.role {
+                    Roles::Goalkeeper => (),
+                    Roles::Defender => {
+                        is_on_its_role = true;
+                    },
+                    Roles::Midfielder => (),
+                    Roles::Attacker => (),
+                }
             },
             Roles::Midfielder => {
-                assert(player.midfielder.is_none(), 'Midfieldr already placed');
+                assert(player.midfielder.is_none(), 'Midfielder already placed');
                 player.midfielder = Option::Some(Placement::Side(card_id));
+                match card.role {
+                    Roles::Goalkeeper => (),
+                    Roles::Defender => (),
+                    Roles::Midfielder => {
+                        is_on_its_role = true;
+                    },
+                    Roles::Attacker => (),
+                }
             },
             Roles::Attacker => {
                 assert(player.attacker.is_none(), 'Attacker already placed');
                 player.attacker = Option::Some(Placement::Side(card_id));
+                match card.role {
+                    Roles::Goalkeeper => (),
+                    Roles::Defender => (),
+                    Roles::Midfielder => (),
+                    Roles::Attacker => {
+                        is_on_its_role = true;
+                    },
+                }
             }
         }
```

Comment:
WDYT of that ?
```rs
        let is_on_its_role = match position {
              Roles::Goalkeeper => {
                  assert(player.goalkeeper.is_none(), 'Goalkeeper already placed');
                  player.goalkeeper = Option::Some(Placement::Side(card_id));
                card.role == Roles::Goalkeeper
              },
              Roles::Defender => {
                  assert(player.defender.is_none(), 'Defender already placed');
                  player.defender = Option::Some(Placement::Side(card_id));
                card.role == Roles::Defender
              },
              Roles::Midfielder => {
                  assert(player.midfielder.is_none(), 'Midfielder already placed');
                  player.midfielder = Option::Some(Placement::Side(card_id));
                card.role == Roles::Midfielder
              },
              Roles::Attacker => {
                  assert(player.attacker.is_none(), 'Attacker already placed');
                  player.attacker = Option::Some(Placement::Side(card_id));
                card.role == Roles::Attacker
              }
        };
```

Reply:
Waaaaayy better hahaha. I encountered a compiler error and concluded it was from the equality test. I thought it was a Cairo limitation with structs not implementing PartialEq, which is why I turned to the match statement.